dkp/dkp.py

```diff
@@ -346,21 +403,21 @@ def backup(
                 fs_copy(bind, dest_path, follow_symlinks=True)
 
         # copy env files
-        file: Path
-        for file in info.work_dir.iterdir():
-            if (
-                file.name != ".env" and file.suffix != ".env"
-            ) or file.resolve().is_dir():
-                continue
-            env_file = project_dir / file.name
-            env_file.parent.mkdir(exist_ok=True, parents=True)
-            print(
-                "found env file",
-                file.name,
-                "- copying to",
-                env_file.relative_to(work_dir),
-            )
-            fs_copy(info.env_file, env_file)
+        if env_files == []:
+            file: Path
+            for file in info.work_dir.iterdir():
+                if (
+                    file.name != ".env" and file.suffix != ".env"
+                ) or file.resolve().is_dir():
+                    continue
+                copy_env_file(file, project_dir / file.name, Path(work_dir))
+        else:
+            env_file: Path
+            for env_file in env_files:
+                copy_env_file(
+                    env_file,
+                    project_dir / env_file.name,
+                    Path(work_dir))
```

Comment:
I afraid it means that `--env-file` and default behaviour are mutually exclusive. Shall we consider combining them together?

My suggestion is to iterate over all env_files (as in second branch) plus on top iterate over default env files in work dir (as in first branch).

But I am open for discussion. If there are strong reasons to keep totally different behaviour, then let's just reflect it in documentation.

Reply:
> My suggestion is to iterate over all env_files (as in second branch) plus on top iterate over default env files in work dir (as in first branch).

I started along this line, with the exception that when at least one env file is specified via `--env-file`, there is no need to copy `.env`. (If you specify at least one `--env-file` to `docker-compose`, it won't use your `.env` file.)

Then I realized that my implementation doesn't handle two cases well.

1.  It doesn't handle it well when `--env-file` is passed, and there are also `*.env` files used by the `env_file` attribute of the compose file. This is related to your remark.

2.  dkp so far copied all env files into the target root directory (`work_dir`) inside the package. This was OK before I added the `--env-file` option because all files were copied from the project root directory (`info.work_dir`). In short: `cp <project_dir>/<env-file-name> <target_project_dir>/<env-file-name>` worked.

    But `--env-file` made it possible to specify files at other locations, not in the root directory of the project. These cases were not handled well in my commit. Here is how we should handle these cases:

    * (a) The env file is outside the project directory.
        - Example: `docker compose --env-file ../my.env`
        - In this case we should skip the env file (just like dkp skips bindigs outside the project directory).
    * (b) The env file is inside a subdirectory in the project directory (maybe multiple directories deep).
        - Example: `docker compose --env-file dir/my.env`
        - In this case we should create the subdirectory in the pack, and copy the env file there.

I modified the code to handle these cases correctly: https://github.com/reddec/dkp/pull/2/commits/fbc2b02f6412daedf5251b0ac47aca6261126b89

It's still not perfect. I documented a remaining limitation in the commit message.